include/seqan3/alphabet/union_alphabet.hpp

```diff
@@ -31,237 +31,411 @@
-struct union_alphabet
+class union_alphabet
 {
-    /* types */
 public:
-    //! The size of the alphabet, i.e. the number of different values it can take.
+    //!\brief The size of the alphabet, i.e. the number of different values it can take.
     static constexpr size_t value_size = (alphabet_types::value_size + ... + first_alphabet_type::value_size);
-    //! the type of the alphabet when converted to char (e.g. via @link to_char @endlink)
-    using char_type = typename first_alphabet_type::char_type;
-    //! the type of the alphabet when represented as a number (e.g. via @link to_rank @endlink)
-    using rank_type = decltype(detail::at_least_type<value_size>());
 
-protected:
-    using tuple_t = std::tuple<first_alphabet_type, alphabet_types...>;
-    using index_sequence_t = decltype(std::make_index_sequence<std::tuple_size_v<tuple_t>>{});
-    static constexpr auto sizes = detail::alphabet_sizes<tuple_t>(index_sequence_t{});
-    static constexpr auto alphabets = detail::alphabet_objects<tuple_t>(index_sequence_t{});
+    //!\brief The type of the alphabet when converted to char (e.g. via \link to_char \endlink)
+    using char_type = typename first_alphabet_type::char_type;
 
-public:
-    //! internal value
-    rank_type value;
+    //!\brief The type of the alphabet when represented as a number (e.g. via \link to_rank \endlink)
+    using rank_type = detail::min_viable_uint_t<value_size>;
+
+    /*!\brief The type used to assign a value from one of the base alphabets
+     * during copy construction or copy assignment.
+     *
+     * For example used by
+     * union_alphabet::union_alphabet(const variant_type & alphabet) or
+     * union_alphabet::operator=(const variant_type & alphabet)
+     */
+    using variant_type = std::variant<first_alphabet_type, alphabet_types...>;
+
+    //!\name Default constructors
+    //!\{
+    constexpr union_alphabet() = default;
+    constexpr union_alphabet(union_alphabet const &) = default;
+    constexpr union_alphabet(union_alphabet &&) = default;
+    //!\}
+
+    //!\name Default assignment operators
+    //!\{
+    constexpr union_alphabet & operator= (union_alphabet const &) = default;
+    constexpr union_alphabet & operator= (union_alphabet &&) = default;
+    //!\}
+
+    //!\name Conversion constructors
+    //\{
+    /*!\brief Construction via a value of the base alphabets
+     *
+     * ```cpp
+     *     union_alphabet<dna4, gap> letter1{dna4::C}; // or
+     *     union_alphabet<dna4, gap> letter2 = gap::GAP;
+     * ```
+     */
+    constexpr union_alphabet(variant_type const & alphabet) :
+        _value{from_base_(alphabet)}
+    {}
+    //!\}
+
+    //!\name Conversion assignment operators
+    //!\{
+    /*!\brief Assignment via a value of the base alphabets
+     *
+     * ```cpp
+     *     union_alphabet<dna4, gap> letter1{};
+     *     letter1 = gap::GAP;
+     * ```
+     */
+    constexpr union_alphabet & operator= (variant_type const & alphabet)
```

Comment:
We hadn't those utility when I wrote this. So I naturally used the stuff from the standard library.

How does this handle the case, if I have the same time type twice in the list?

With variants there I can express:

```c++
constexpr alphabet_t letter0{variant_t{std::in_place_index_t<0>{}, dna4::A}};
constexpr alphabet_t letter4{variant_t{std::in_place_index_t<1>{}, dna4::A}};
EXPECT_EQ(letter0.to_rank(), 0);
EXPECT_EQ(letter4.to_rank(), 4);
```